src/fdosecrets/widgets/SettingsModels.cpp

```diff
@@ -0,0 +1,396 @@
+        default:
+            return {};
+        }
+    }
+
+    QVariant SettingsDatabaseModel::dataForName(DatabaseWidget* db, int role) const
+    {
+        switch (role) {
+        case Qt::DisplayRole: {
+            QFileInfo fi(db->database()->filePath());
+            return fi.fileName();
+        }
+        case Qt::ToolTipRole:
+            return db->database()->filePath();
+        default:
+            return {};
+        }
+    }
+
+    QVariant SettingsDatabaseModel::dataForExposedGroup(DatabaseWidget* dbWidget, int role)
+    {
+        if (dbWidget->isLocked()) {
+            switch (role) {
+            case Qt::DisplayRole:
+                return tr("Unlock to show");
+            case Qt::DecorationRole:
+                return filePath()->icon(QStringLiteral("apps"), QStringLiteral("object-locked"), true);
+            case Qt::FontRole: {
+                QFont font;
+                font.setItalic(true);
+                return font;
+            }
+            default:
+                return {};
+            }
+        }
+        auto db = dbWidget->database();
+        auto group = db->rootGroup()->findGroupByUuid(FdoSecrets::settings()->exposedGroup(db));
+        if (group) {
+            switch (role) {
+            case Qt::DisplayRole:
+                return group->name();
+            case Qt::DecorationRole:
+                return group->isExpired() ? databaseIcons()->iconPixmap(DatabaseIcons::ExpiredIconIndex)
+                                          : group->iconScaledPixmap();
+            case Qt::FontRole:
+                if (group->isExpired()) {
+                    QFont font;
+                    font.setStrikeOut(true);
+                    return font;
+                } else {
+                    return {};
+                }
+            default:
+                return {};
+            }
+        } else {
+            switch (role) {
+            case Qt::DisplayRole:
+                return tr("None");
+            case Qt::DecorationRole:
+                return filePath()->icon(QStringLiteral("apps"), QStringLiteral("paint-none"), true);
+            default:
+                return {};
+            }
+        }
+    }
+
+    QVariant SettingsDatabaseModel::dataForManage(DatabaseWidget* db, int role) const
+    {
+        switch (role) {
+        case Qt::EditRole:
+            return QVariant::fromValue(db);
+        default:
+            return {};
+        }
+    }
+
+    void SettingsDatabaseModel::populateModel()
+    {
+        beginResetModel();
+
+        m_dbs.clear();
+
+        if (m_dbTabs) {
+            // Add existing database tabs
+            for (int idx = 0; idx != m_dbTabs->count(); ++idx) {
+                auto dbWidget = m_dbTabs->databaseWidgetFromIndex(idx);
+                databaseAdded(dbWidget, false);
+            }
+            // connect signals
+            connect(m_dbTabs, &DatabaseTabWidget::databaseOpened, this, [this](DatabaseWidget* db) {
```

Comment:
I do not like this because it is the exact opposite of how signal/slots should be used. The Settings Model should be created, then the signals connected from the parent. The settings model should have no knowledge of the db tab widget.

I keep it as-is for now, but this would be a good upgrade for 2.6.0 to break some hard dependencies.

Reply:
I'll keep that in mind. In general, I'm also thinking about removing dependencies on db tab widget.

(Or even better, don't depend on any GUI classes at all. But that's quite difficult for now, given the db widget is the only one keeping the filename and lock state).

Reply:
The filename is located in the Database class. You can determine if its locked by checking to see if it has a key associated with it.